src/clj_puppetdb/http.clj

```diff
@@ -24,47 +24,86 @@
   `puppetlabs.ssl-utils.core/pems->ssl-context` function."
   [:ssl-cert :ssl-key :ssl-ca-cert])
 
+(defn- make-client-common
+  [^String host opts]
+  (let [opts (assoc opts :as :stream)
+        info {:host host}
+        info (if-let [ssl-context (:ssl-context opts)]
+               (assoc info :ssl-context ssl-context)
+               info)]
+    (fn
+      ; This arity-overloaded variant of the function is the one called by the `GET` function.
+      ; The `this` parameter refers under normal circumstances to this very function and it is
+      ; needed to be able to call another arity-overloaded variant of the function from the
+      ; function body. When the `this` parameter refers to a different function then it
+      ; effectively overrides the other arity-overloaded variants of the function.
+      ([this path params]
+       (let [query (-> (StringBuilder. host)
+                       (.append path))]
+         (if-not (empty? params)
+           (-> query
+               (.append \?)
+               (.append (-> params
+                            q/params->json
+                            map->query))))
+         ; now call the single argument variant of the function to execute the query
+         (this (.toString query))))
+
+      ; This arity-overloaded variant of the function is the one which actually executes
+      ; the PDB query.
+      ([query]
+        #_(println "GET:" query)                                 ;; uncomment this to watch queries
```

Comment:
IMHO, it should be possible to use standard logging mechanism with different log levels. This would probably be in DEBUG log level.

Take a look at pe-rbac-service for inspiration.